### ISSUE TITLE
LangChain's Prompt  is like a holy shit design. So many errors when i uesd it in AgentExecutor.Could you improve it ?

### ISSUE THREAD

**Issue by imagebody** [langchain-ai/langchain#20769]
### Checked other resources

- [X] I added a very descriptive title to this issue.
- [X] I searched the LangChain documentation with the integrated search.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I am sure that this is a bug in LangChain rather than my code.
- [X] The bug is not resolved by updating to the latest stable version of LangChain (or the specific integration package).

### Example Code

    prompt_template = PromptTemplate.from_template(
        PROMPT_TEMPLATE
    )

    missing_vars = {"tools", "tool_names", "agent_scratchpad"}.difference(
        prompt_template.input_variables
    )
    if missing_vars:
        raise ValueError(f"Prompt missing required variables: {missing_vars}")
    from langchain.agents.format_scratchpad import format_log_to_str
    prompt = prompt_template.format(
        tools=tools_renderer(list(tools)),
        tool_names=", ".join([t.name for t in tools]),
    )
    
    

### Description

    prompt_template = PromptTemplate.from_template(
        PROMPT_TEMPLATE
    )

    missing_vars = {"tools", "tool_names", "agent_scratchpad"}.difference(
        prompt_template.input_variables
    )
    if missing_vars:
        raise ValueError(f"Prompt missing required variables: {missing_vars}")
    from langchain.agents.format_scratchpad import format_log_to_str
    prompt = prompt_template.format(
        tools=tools_renderer(list(tools)),
        tool_names=", ".join([t.name for t in tools]),
    )

### System Info

    prompt_template = PromptTemplate.from_template(
        PROMPT_TEMPLATE
    )

    missing_vars = {"tools", "tool_names", "agent_scratchpad"}.difference(
        prompt_template.input_variables
    )
    if missing_vars:
        raise ValueError(f"Prompt missing required variables: {missing_vars}")
    from langchain.agents.format_scratchpad import format_log_to_str
    prompt = prompt_template.format(
        tools=tools_renderer(list(tools)),
        tool_names=", ".join([t.name for t in tools]),
    )

**Comment by coolmian** [comment]
I completely agree that the JSON prompt word I used in the prompt word will be incorrectly recognized as a variable by the f-string formatter, so the error requires me to fill in the variable, which is simply bullshit

**Comment by coolmian** [comment]
**I solved this problem by using a class without a formatter**

```python
from langchain_core.prompts import ChatPromptTemplate, MessagesPlaceholder
from langchain_core.messages import AIMessage
from langchain_openai import ChatOpenAI
chat = ChatOpenAI()
role_play_prompt = ChatPromptTemplate.from_messages(
    [
        (
            "system", "you are an assistant",
        ),
        # here use HumanMessage/AIMessage/SystemMessage instead of string tuple likes ("system", "xxxxx")
        AIMessage(content="{'reply': 'ok'}"),
        MessagesPlaceholder(variable_name="messages"),
    ]
)

chain = role_play_prompt | chat
chain.invoke({"messages": [{"role": "user", "content": "hello"}]})

```

**Comment by ekkkkki** [comment]
Same problem. It hasn't been fixed for months.
